src/lib/include/x86_vector_triple_basemath.h

```diff
@@ -0,0 +1,139 @@
+#ifndef DISTOPIA_X86_VECTOR_TRIPLE_BASEMATH_H
+#define DISTOPIA_X86_VECTOR_TRIPLE_BASEMATH_H
+
+#include "arch_config.h"
+
+#ifdef DISTOPIA_X86_SSE4_1
+
+#include <cfloat>
+#include <immintrin.h>
+#include <type_traits>
+
+#include "compiler_hints.h"
+#include "distopia_type_traits.h"
+#include "ops.h"
+#include "vector_triple.h"
+#include "x86_basemath.h"
+#include "x86_tgintrin.h"
+#include "x86_vectors.h"
+
+namespace {
+
+// basemath for VectorTriples
+template <typename VectorT>
+inline VectorTriple<VectorT> Abs(VectorTriple<VectorT> source) {
+  return VectorTriple<VectorT>(Abs(source.x), Abs(source.y), Abs(source.z));
+}
+
+template <typename VectorT>
+inline VectorTriple<VectorT> NearbyInt(VectorTriple<VectorT> source) {
+  return VectorTriple<VectorT>(Nearbyint(source.x), Nearbyint(source.y),
+                               Nearbyint(source.z));
+}
+
+template <typename VectorT>
+inline VectorTriple<VectorT> Sqrt(VectorTriple<VectorT> source) {
+  return VectorTriple<VectorT>(Sqrt(source.x), Sqrt(source.y), Sqrt(source.z));
+}
+
+#ifdef DISTOPIA_X86_AVX2_FMA
+// fast version with FMA intrinsics.
+template <typename VectorT>
+inline VectorTriple<VectorT> FastMulAdd(VectorTriple<VectorT> a,
```

Comment:
I kinda hate calling stuff "Fast". This is fused right? Call it that

Reply:
Ill change it.

Reply:
@richardjgowers @hmacdope It’s not fused. ‘Fused multiply-add’ has a specific technical meaning: it’s equivalent to performing the multiplication in infinite precision and only rounding after the addition. The result is different from, and more accurate than, doing multiplication and addition separately. I called it ‘fast’ because it automatically chooses the faster implementation depending on the architecture, even if it results in less accurate results. We can discuss whether that’s a good name, but I’m reverting this particular change because it’s incorrect.

Reply:
Btw, when I call things _fast_ that’s short for 
> Caution! This code has potentially-surprising behavior that permits more optimizations. Make sure you’re using it safely.

In this case the surprising behavior is irreproducibility between architectures. `FastMin` gives results that are different from `std::fmin` and also handles some NaN cases incorrectly.

I wouldn’t put ‘fast’ in the name of a function without surprising behavior. I agree that’d be ugly.